dsgrid/project.py

```diff
@@ -183,32 +170,26 @@ def process_query(self, context: QueryContext, cached_project_mapped_datasets_di
         dim_columns = sorted(dim_columns)
         time_columns = sorted(time_columns)
         expected_columns = time_columns + pivoted_columns_sorted + dim_columns
-        for i, df in enumerate(dfs):
+
+        datasets = {}
+        for dataset_id, path in df_filenames.items():
+            df = read_dataframe(path)
             remaining = sorted(set(df.columns).difference(expected_columns))
             final_columns = expected_columns + remaining
             missing = pivoted_columns.difference(df.columns)
             for column in missing:
                 df = df.withColumn(column, F.lit(None).cast(DoubleType()))
-            df = df.select(*final_columns)
-            dfs[i] = df
-
-        main_df = dfs[0]
-        if len(dfs) > 1:
-            for df in dfs[1:]:
-                if df.columns != main_df.columns:
-                    raise Exception(
-                        f"BUG: dataset columns must match. main={main_df.columns} new={df.columns} "
-                        f"diff={set(main_df.columns).symmetric_difference(df.columns)}"
-                    )
-                main_df = main_df.union(df)
+            datasets[dataset_id] = DatasetExpressionHandler(
+                df.select(*final_columns), time_columns + dim_columns, pivoted_columns_sorted
```

Comment:
I think you can just pass in df into DatasetExpressionHandler here, ordering needs to be enforced at the final output.

Reply:
When you perform DataFrame.union the dataframes have to have identical column order. This ensure that happens. Or am I missing something?

Reply:
I think this just turns it into an object. The ordering here doesn't matter I think. This comment goes with the next comment

Reply:
Is the issue that I'm putting dimension columns after pivoted columns in project.py but then before pivoted columns in the expression handler? That can be fixed. The main thing is that the columns in all datasets have to be in the same order when you perform the union, otherwise you get garbage.

Reply:
I fixed column ordering in the dataset expression code - that shouldn't change order. I'm very tempted to stop worrying about order in our processing code until the end, however. Every time you join, Spark puts the join columns first.

Reply:
Yes, I agree, for union, we could just have a utility func that attempts to make the cols in two df the same, and the final ordering can be a last step.